src/runtime/relax_vm/paged_kv_cache.cc

```diff
@@ -0,0 +1,633 @@
+   * \note Since page table is a ragged data structure, we represent it
+   * in CSR format (which uses indptr and values below) on device.
+   */
+  NDArray page_table_indptr_device_;
+  /*! \brief The page table value array on device. */
+  NDArray page_table_values_device_;
+  /*!
+   * \brief The array storing "the number of used slots in the last page
+   * of each sequence" on device. Its values range in (0, page_size_].
+   */
+  NDArray last_page_offset_device_;
+  /*!
+   * \brief The append_length indptr array on device.
+   * \note Since the Q/K/V data may have raggedness in terms of lengths,
+   * we represent the the append lengths in CSR format.
+   */
+  NDArray cur_append_length_indptr_device_;
+  /*!
+   * \brief The corresponding sequence id for each position along the
+   * length dimension of K/V data. It is used for efficient computation.
+   */
+  NDArray cur_pos2seqid_device_;
+
+ public:
+  /*! \brief Constructor. Take the cache configuration and initialize the NDArrays. */
+  explicit PagedAttentionKVCacheObj(int64_t page_size, int64_t num_layers, int64_t num_heads,
+                                    int64_t head_dim, int64_t reserved_num_seqs,
+                                    int64_t reserved_num_pages, DLDataType dtype, DLDevice device)
+      : page_size_(page_size), num_layers_(num_layers), num_heads_(num_heads), head_dim_(head_dim) {
+    pages_ = NDArray::Empty({reserved_num_pages, num_layers, 2, num_heads, page_size, head_dim},
+                            dtype, device);
+    page_table_indptr_device_ = NDArray::Empty({reserved_num_seqs + 1}, dtype_aux_, device);
+    page_table_values_device_ = NDArray::Empty({reserved_num_pages}, dtype_aux_, device);
+    last_page_offset_device_ = NDArray::Empty({reserved_num_seqs}, dtype_aux_, device);
+    cur_append_length_indptr_device_ = NDArray::Empty({reserved_num_seqs + 1}, dtype_aux_, device);
+    cur_pos2seqid_device_ = NDArray::Empty({reserved_num_pages * page_size}, dtype_aux_, device);
+  }
+
+  /*!
+   * \brief Given a sequence id and a required extra length, allocate new
+   * new pages for the sequence until the total capacity can cover the
+   * current sequence length plus the required extra length.
+   * \note By reserving extra length for the sequence, the subsequent appended
+   * K/V data for the sequence must have the same length as the input length here.
+   * \param seq_id The id of the sequence to process.
+   * \param extra_length The extra length to reserve for the sequence.
+   */
+  void ReserveExtraLengthForAppend(int64_t seq_id, int64_t extra_length) {
+    CHECK_GE(seq_id, 0) << "Input sequence id should be positive";
+    CHECK_LE(seq_id, num_total_seqs_)
+        << "Invalid input sequence id " << seq_id
+        << ", which is larger than the current total number of sequences.";
+    CHECK_GT(extra_length, 0) << "The input length should be positive.";
+
+    if (seq_id == num_total_seqs_) {
```

Comment:
I think it is useful to have a separate function to grow the sequence to keep the concern of concurrent sequence growing and seq len growing separately